File: v2/src/shared/microsoft.js
/**
 * Microsoft Graph API integration
 * Uses client credentials flow (app-only auth) for tenant-wide access
 *
 * Required Azure AD App permissions (Application, not Delegated):
 * - Application.Read.All
 * - Directory.Read.All
 * - AuditLog.Read.All (optional, for consent history)
 */

const { SecretsManagerClient, GetSecretValueCommand } = require('@aws-sdk/client-secrets-manager');

const secretsClient = new SecretsManagerClient({});

// Token cache: { tenantId: { token, expiry } }
const tokenCache = new Map();

/**
 * Get client secret from Secrets Manager
 */
async function getClientSecret(secretArn) {
  const result = await secretsClient.send(new GetSecretValueCommand({
    SecretId: secretArn,
  }));
  return result.SecretString;
}

/**
 * Get app-only access token using client credentials flow
 * This allows tenant-wide access without user interaction
 */
exports.getAppOnlyToken = async (tenantId, clientId, clientSecretArn) => {
  // Check cache first
  const cached = tokenCache.get(tenantId);
  if (cached && cached.expiry > Date.now() + 60000) { // 1 min buffer
    return cached.token;
  }

  const clientSecret = await getClientSecret(clientSecretArn);

  const response = await fetch(
    `https://login.microsoftonline.com/${tenantId}/oauth2/v2.0/token`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body: new URLSearchParams({
        client_id: clientId,
        client_secret: clientSecret,
        scope: 'https://graph.microsoft.com/.default',
        grant_type: 'client_credentials',
      }),
    }
  );

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Failed to get app-only token: ${error}`);
  }

  const data = await response.json();
  const token = data.access_token;
  const expiry = Date.now() + (data.expires_in * 1000);

  // Cache the token
  tokenCache.set(tenantId, { token, expiry });

  return token;
};

/**
 * Validate connection credentials by attempting to get a token and org info
 */
exports.validateConnection = async (tenantId, clientId, clientSecretArn) => {
  try {
    const token = await exports.getAppOnlyToken(tenantId, clientId, clientSecretArn);
    const org = await exports.getOrganization(token);

    return {
      valid: true,
      tenantId: org.id,
      tenantName: org.displayName,
      verifiedDomains: org.verifiedDomains?.map(d => d.name) || [],
    };
  } catch (error) {
    return {
      valid: false,
      error: error.message,
    };
  }
};

/**
 * Make authenticated Graph API request
 */
async function graphRequest(accessToken, endpoint, options = {}) {
  const url = endpoint.startsWith('http')
    ? endpoint
    : `https://graph.microsoft.com/v1.0${endpoint}`;

  const response = await fetch(url, {
    ...options,
    headers: {
      Authorization: `Bearer ${accessToken}`,
      'Content-Type': 'application/json',
      ...options.headers,
    },
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Graph API error (${response.status}): ${error}`);
  }

  return response.json();
}

/**
 * Get organization info
 */
exports.getOrganization = async (accessToken) => {
  const result = await graphRequest(accessToken, '/organization');
  return result.value?.[0];
};

/**
 * List all service principals (OAuth/Enterprise apps)
 */
exports.listServicePrincipals = async (accessToken) => {
  const apps = [];
  let nextLink = '/servicePrincipals?$select=id,appId,displayName,appDisplayName,publisherName,homepage,replyUrls,servicePrincipalType,tags,accountEnabled,createdDateTime,appOwnerOrganizationId&$top=100';

  while (nextLink) {
    const result = await graphRequest(accessToken, nextLink);
    apps.push(...(result.value || []));
    nextLink = result['@odata.nextLink'];
  }

  return apps;
};

/**
 * Get OAuth2 permission grants (delegated permissions that have been consented)
 */
exports.listOAuth2Grants = async (accessToken) => {
  const grants = [];
  let nextLink = '/oauth2PermissionGrants?$top=100';

  while (nextLink) {
    const result = await graphRequest(accessToken, nextLink);
    grants.push(...(result.value || []));
    nextLink = result['@odata.nextLink'];
  }

  return grants;
};

/**
 * Get app role assignments for a service principal (application permissions)
 */
exports.listAppRoleAssignments = async (accessToken, servicePrincipalId) => {
  try {
    const result = await graphRequest(
      accessToken,
      `/servicePrincipals/${servicePrincipalId}/appRoleAssignments`
    );
    return result.value || [];
  } catch (e) {
    return [];
  }
};

/**
 * Get all app role assignments in the tenant
 */
exports.listAllAppRoleAssignments = async (accessToken, servicePrincipals) => {
  const assignments = new Map();

  // Get app role assignments for each service principal
  // This shows what application permissions each app has
  for (const sp of servicePrincipals) {
    const roles = await exports.listAppRoleAssignments(accessToken, sp.id);
    if (roles.length > 0) {
      assignments.set(sp.id, roles);
    }
  }

  return assignments;
};

/**
 * Full OAuth App Audit scan
 * Returns structured data about all OAuth apps and their permissions
 */
exports.scanOAuthApps = async (accessToken) => {
  // Get all service principals
  const servicePrincipals = await exports.listServicePrincipals(accessToken);

  // Get all delegated permission grants
  const grants = await exports.listOAuth2Grants(accessToken);

  // Build a map of grants by service principal ID
  const grantsByPrincipal = {};
  for (const grant of grants) {
    const clientId = grant.clientId;
    if (!grantsByPrincipal[clientId]) {
      grantsByPrincipal[clientId] = [];
    }
    grantsByPrincipal[clientId].push({
      scope: grant.scope,
      consentType: grant.consentType,
      principalId: grant.principalId,
    });
  }

  // Filter to third-party apps and build audit results
  const auditedApps = servicePrincipals
    .filter(sp => {
      // Include apps that are:
      // 1. Application type (not managed identity, etc.)
      // 2. Third-party (different owner org) OR have permission grants
      return sp.servicePrincipalType === 'Application' &&
        (sp.appOwnerOrganizationId !== sp.appId || grantsByPrincipal[sp.id]);
    })
    .map(sp => {
      const delegatedGrants = grantsByPrincipal[sp.id] || [];
      const delegatedScopes = [...new Set(
        delegatedGrants.flatMap(g => g.scope?.split(' ') || []).filter(Boolean)
      )];

      const isFirstParty = isFirstPartyApp(sp.appOwnerOrganizationId);

      return {
        appId: sp.appId,
        servicePrincipalId: sp.id,
        displayName: sp.appDisplayName || sp.displayName,
        publisher: sp.publisherName || (isFirstParty ? 'Microsoft' : 'Unknown'),
        publisherOrgId: sp.appOwnerOrganizationId,
        isFirstParty,
        enabled: sp.accountEnabled,
        createdAt: sp.createdDateTime,
        homepage: sp.homepage,
        delegatedPermissions: delegatedScopes,
        consentType: delegatedGrants.some(g => g.consentType === 'AllPrincipals')
          ? 'admin_consent'
          : delegatedGrants.length > 0
            ? 'user_consent'
            : 'none',
        userCount: new Set(delegatedGrants.map(g => g.principalId).filter(Boolean)).size,
        riskLevel: calculateRiskLevel(delegatedScopes, isFirstParty),
        riskFactors: getRiskFactors(delegatedScopes, sp),
      };
    });

  // Sort by risk level
  const riskOrder = { high: 0, medium: 1, low: 2 };
  auditedApps.sort((a, b) => riskOrder[a.riskLevel] - riskOrder[b.riskLevel]);

  return {
    totalApps: auditedApps.length,
    summary: {
      highRisk: auditedApps.filter(a => a.riskLevel === 'high').length,
      mediumRisk: auditedApps.filter(a => a.riskLevel === 'medium').length,
      lowRisk: auditedApps.filter(a => a.riskLevel === 'low').length,
      thirdParty: auditedApps.filter(a => !a.isFirstParty).length,
      firstParty: auditedApps.filter(a => a.isFirstParty).length,
    },
    apps: auditedApps,
    scannedAt: new Date().toISOString(),
  };
};

/**
 * Check if app is Microsoft first-party
 */
function isFirstPartyApp(ownerOrgId) {
  const microsoftOrgIds = [
    'f8cdef31-a31e-4b4a-93e4-5f571e91255a', // Microsoft Services
    '72f988bf-86f1-41af-91ab-2d7cd011db47', // Microsoft Corp
  ];
  return microsoftOrgIds.includes(ownerOrgId);
}

/**
 * Calculate risk level based on permissions
 */
function calculateRiskLevel(scopes, isFirstParty) {
  // First-party Microsoft apps are generally lower risk
  if (isFirstParty && scopes.length === 0) return 'low';

  const highRiskScopes = [
    'Mail.ReadWrite', 'Mail.ReadWrite.All',
    'Mail.Send', 'Mail.Send.All',
    'Files.ReadWrite.All',
    'Directory.ReadWrite.All',
    'User.ReadWrite.All',
    'Application.ReadWrite.All',
    'RoleManagement.ReadWrite.Directory',
    'Sites.ReadWrite.All',
    'Group.ReadWrite.All',
  ];

  const mediumRiskScopes = [
    'Mail.Read', 'Mail.Read.All',
    'Files.Read.All',
    'Calendars.ReadWrite', 'Calendars.ReadWrite.All',
    'Contacts.ReadWrite', 'Contacts.ReadWrite.All',
    'Directory.Read.All',
    'User.Read.All',
    'Sites.Read.All',
    'Group.Read.All',
  ];

  const hasHighRisk = scopes.some(s => highRiskScopes.includes(s));
  const hasMediumRisk = scopes.some(s => mediumRiskScopes.includes(s));

  if (hasHighRisk) return 'high';
  if (hasMediumRisk) return 'medium';
  return 'low';
}

/**
 * Get specific risk factors for an app
 */
function getRiskFactors(scopes, sp) {
  const factors = [];

  // Check for risky permissions
  if (scopes.some(s => s.includes('Mail.Send'))) {
    factors.push({ type: 'permission', severity: 'high', description: 'Can send email as users' });
  }
  if (scopes.some(s => s.includes('Mail.ReadWrite'))) {
    factors.push({ type: 'permission', severity: 'high', description: 'Can read and modify email' });
  }
  if (scopes.some(s => s.includes('Files.ReadWrite.All'))) {
    factors.push({ type: 'permission', severity: 'high', description: 'Can access all files in OneDrive/SharePoint' });
  }
  if (scopes.some(s => s.includes('Directory.ReadWrite'))) {
    factors.push({ type: 'permission', severity: 'high', description: 'Can modify directory objects' });
  }

  // Check for unknown publisher
  if (!sp.publisherName && !isFirstPartyApp(sp.appOwnerOrganizationId)) {
    factors.push({ type: 'publisher', severity: 'medium', description: 'Unknown publisher' });
  }

  // Check if app is disabled (might indicate it was found malicious)
  if (!sp.accountEnabled) {
    factors.push({ type: 'status', severity: 'info', description: 'App is disabled' });
  }

  return factors;
}

File: desktop/src-tauri/src/platforms/microsoft.rs
//! Microsoft platform adapter
//!
//! Handles OAuth and app discovery for Microsoft personal accounts.
//! Uses Microsoft Graph API to list third-party apps.

use super::{calculate_risk_level, ConnectedAccount, DiscoveredApp, Platform, PlatformError};
use serde::Deserialize;

// OAuth configuration
const CLIENT_ID: &str = "YOUR_MICROSOFT_CLIENT_ID";
const REDIRECT_URI: &str = "http://localhost:8742/callback/microsoft";
const AUTH_URL: &str = "https://login.microsoftonline.com/consumers/oauth2/v2.0/authorize";
const TOKEN_URL: &str = "https://login.microsoftonline.com/consumers/oauth2/v2.0/token";

// Scopes for personal Microsoft accounts
const SCOPES: &[&str] = &[
    "openid",
    "email",
    "profile",
    "offline_access",
    "User.Read",
];

pub struct MicrosoftPlatform {
    client: reqwest::Client,
}

impl MicrosoftPlatform {
    pub fn new() -> Self {
        Self {
            client: reqwest::Client::new(),
        }
    }
}

#[derive(Debug, Deserialize)]
struct TokenResponse {
    access_token: String,
    refresh_token: Option<String>,
    expires_in: Option<i64>,
    token_type: String,
}

#[derive(Debug, Deserialize)]
struct UserInfo {
    #[serde(rename = "userPrincipalName")]
    user_principal_name: Option<String>,
    mail: Option<String>,
    #[serde(rename = "displayName")]
    display_name: Option<String>,
}

#[derive(Debug, Deserialize)]
struct OAuthPermissionGrant {
    id: String,
    #[serde(rename = "clientId")]
    client_id: String,
    #[serde(rename = "consentType")]
    consent_type: Option<String>,
    scope: Option<String>,
    #[serde(rename = "principalId")]
    principal_id: Option<String>,
}

#[derive(Debug, Deserialize)]
struct ServicePrincipal {
    id: String,
    #[serde(rename = "appId")]
    app_id: String,
    #[serde(rename = "displayName")]
    display_name: Option<String>,
    #[serde(rename = "appDisplayName")]
    app_display_name: Option<String>,
    #[serde(rename = "publisherName")]
    publisher_name: Option<String>,
    homepage: Option<String>,
    #[serde(rename = "appOwnerOrganizationId")]
    app_owner_org_id: Option<String>,
}

#[derive(Debug, Deserialize)]
struct GraphResponse<T> {
    value: Vec<T>,
    #[serde(rename = "@odata.nextLink")]
    next_link: Option<String>,
}

// Microsoft first-party org IDs
const MICROSOFT_ORG_IDS: &[&str] = &[
    "f8cdef31-a31e-4b4a-93e4-5f571e91255a",
    "72f988bf-86f1-41af-91ab-2d7cd011db47",
];

impl Platform for MicrosoftPlatform {
    fn get_auth_url(&self) -> Result<String, PlatformError> {
        let scopes = SCOPES.join(" ");
        let url = format!(
            "{}?client_id={}&redirect_uri={}&response_type=code&scope={}&response_mode=query",
            AUTH_URL,
            CLIENT_ID,
            urlencoding::encode(REDIRECT_URI),
            urlencoding::encode(&scopes)
        );
        Ok(url)
    }

    async fn exchange_code(&self, code: &str) -> Result<ConnectedAccount, PlatformError> {
        let params = [
            ("code", code),
            ("client_id", CLIENT_ID),
            ("redirect_uri", REDIRECT_URI),
            ("grant_type", "authorization_code"),
            ("scope", &SCOPES.join(" ")),
        ];

        let response: TokenResponse = self
            .client
            .post(TOKEN_URL)
            .form(&params)
            .send()
            .await?
            .json()
            .await?;

        // Get user info
        let user_info: UserInfo = self
            .client
            .get("https://graph.microsoft.com/v1.0/me")
            .bearer_auth(&response.access_token)
            .send()
            .await?
            .json()
            .await?;

        let email = user_info
            .mail
            .or(user_info.user_principal_name)
            .unwrap_or_default();

        Ok(ConnectedAccount {
            platform: "microsoft".to_string(),
            email,
            display_name: user_info.display_name,
            access_token: response.access_token,
            refresh_token: response.refresh_token,
            token_expires_at: response.expires_in.map(|e| {
                chrono::Utc::now()
                    .checked_add_signed(chrono::Duration::seconds(e))
                    .map(|t| t.to_rfc3339())
                    .unwrap_or_default()
            }),
            scopes: SCOPES.iter().map(|s| s.to_string()).collect(),
        })
    }

    async fn refresh_token(&self, refresh_token: &str) -> Result<ConnectedAccount, PlatformError> {
        let params = [
            ("refresh_token", refresh_token),
            ("client_id", CLIENT_ID),
            ("grant_type", "refresh_token"),
            ("scope", &SCOPES.join(" ")),
        ];

        let response: TokenResponse = self
            .client
            .post(TOKEN_URL)
            .form(&params)
            .send()
            .await?
            .json()
            .await?;

        let user_info: UserInfo = self
            .client
            .get("https://graph.microsoft.com/v1.0/me")
            .bearer_auth(&response.access_token)
            .send()
            .await?
            .json()
            .await?;

        let email = user_info
            .mail
            .or(user_info.user_principal_name)
            .unwrap_or_default();

        Ok(ConnectedAccount {
            platform: "microsoft".to_string(),
            email,
            display_name: user_info.display_name,
            access_token: response.access_token,
            refresh_token: response.refresh_token.or(Some(refresh_token.to_string())),
            token_expires_at: response.expires_in.map(|e| {
                chrono::Utc::now()
                    .checked_add_signed(chrono::Duration::seconds(e))
                    .map(|t| t.to_rfc3339())
                    .unwrap_or_default()
            }),
            scopes: SCOPES.iter().map(|s| s.to_string()).collect(),
        })
    }

    async fn scan_apps(&self, access_token: &str) -> Result<Vec<DiscoveredApp>, PlatformError> {
        // For personal Microsoft accounts, we can't use the same Graph APIs
        // as enterprise (no servicePrincipals endpoint for consumers)
        //
        // Instead, direct users to the Microsoft consent management page
        // https://account.live.com/consent/Manage

        let mut apps = Vec::new();

        // Check what we can access with current token
        let me_response = self
            .client
            .get("https://graph.microsoft.com/v1.0/me")
            .bearer_auth(access_token)
            .send()
            .await;

        if me_response.is_ok() {
            // Token is valid - we can at least confirm the connection works
            apps.push(DiscoveredApp {
                app_id: "complens-scanner".to_string(),
                name: "Complens (this app)".to_string(),
                publisher: Some("Complens".to_string()),
                description: Some("This is the Complens app you're using to scan.".to_string()),
                homepage_url: Some("https://complens.ai".to_string()),
                icon_url: None,
                permissions: SCOPES.iter().map(|s| s.to_string()).collect(),
                consent_type: Some("user".to_string()),
                consented_at: None,
                risk_level: "low".to_string(),
                risk_factors: vec![],
                is_first_party: false,
            });
        }

        // Add link to Microsoft's consent management
        apps.push(DiscoveredApp {
            app_id: "microsoft-consent-page".to_string(),
            name: "View all apps at Microsoft".to_string(),
            publisher: Some("Microsoft".to_string()),
            description: Some(
                "Microsoft personal accounts require manual review. \
                 Click to view your connected apps directly on Microsoft."
                    .to_string(),
            ),
            homepage_url: Some("https://account.live.com/consent/Manage".to_string()),
            icon_url: None,
            permissions: vec![],
            consent_type: None,
            consented_at: None,
            risk_level: "info".to_string(),
            risk_factors: vec!["Manual review required".to_string()],
            is_first_party: true,
        });

        Ok(apps)
    }

    async fn revoke_app(
        &self,
        _access_token: &str,
        _app_id: &str,
    ) -> Result<Option<String>, PlatformError> {
        // Microsoft requires manual revocation through their UI
        Ok(Some(
            "https://account.live.com/consent/Manage".to_string(),
        ))
    }
}
